### ISSUE TITLE
Broken ES Module in version 1.1.0

### ISSUE THREAD

**Issue by adrian-branescu** [alumna/reflect#26]
There is a contradiction in the version 1.1.0 of the package.
You state that the entry-point of this package should be treated as an ES Module by setting:
```json
"type": "module"
```
in [package.json](https://github.com/alumna/reflect/blob/master/package.json#L7).

You also set the entry-point of this package to a CommonJS file:
```json
"main": "dist/reflect.cjs.js"
```
in [package.json](https://github.com/alumna/reflect/blob/master/package.json#L5).

The end result is that when you _import_ `@alumna/reflect` you get a broken ES Module. Also, if you try to _require_ it, the Node.js module loader complains:
```bash
reflect.cjs.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
```

The right solution is to use [conditional exports](https://nodejs.org/api/packages.html#conditional-exports) with entries for _import_ and _require_.

**Comment by paulocoghi** [comment]
Thanks, Adrian! Solving it right now :zap:

**Comment by paulocoghi** [comment]
Fix released on v1.1.1. @adrian-branescu, please, when possible, check if it is working again now.

**Comment by adrian-branescu** [comment]
> Fix released on v1.1.1. @adrian-branescu, please, when possible, check if it is working again now.

It complains about the path you've used:
```bash
Error [ERR_INVALID_PACKAGE_TARGET]: Invalid "exports" main target "dist/reflect.es.js" defined in the package config /data/workspace/warpos/build.engage-py.core/node_modules/@alumna/reflect/package.json imported from /data/workspace/warpos/build.engage-py.core/repl; targets must start with "./"
```

I've locally edited the package.json and it seems ok after adding the "./" prefix to the paths.

Also, for the CommonJS version to work, I think you have to use `.cjs` extension instead of `.cjs.js` because it still complains:
```bash
reflect.cjs.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
Instead rename reflect.cjs.js to end in .cjs
```

**Comment by paulocoghi** [comment]
Ok, fixing.

**Comment by paulocoghi** [comment]
Sorry for the delay and thanks for the excellent references! I read more carefully the recommendations for "[Dual CommonJS/ES module packages](https://nodejs.org/api/packages.html#dual-commonjses-module-packages)" _(the next section after the one you linked)_ and I learned a lot from both sections.

When possible, I cordially ask you to test v1.1.3

**Comment by paulocoghi** [comment]
After testing all possibilities, I concluded that the most backwards-compatible strategy is

```json
{
	"exports": {
		"import": "./dist/reflect.mjs",
		"require": "./dist/reflect.cjs",
		"default": "./dist/reflect.mjs"
	},
	"main": "dist/reflect.mjs",
	"module": "dist/reflect.mjs",
	"type": "module",
}
```

---

And the rationale is:

- both ESM and CommonJS load fine on their respective scenarios (this is the most important and most desired)
- conditional exports have precedence over "main" and "module"
- since the package is now a ES module, main points to the `.mjs` file, solving the contradiction
- and, finally, "module" is there just for backward compatibility reasons

This is finally available in v1.1.3 :tada: :smile:
